### PR TITLE
Issue 3438 - Struct ctor with all defaulted parameters should be rejected

### PR DESCRIPTION
**Pull request by AndrejMitrovic**
Fixes https://issues.dlang.org/show_bug.cgi?id=3438

Recreated from: https://github.com/D-Programming-Language/dmd/pull/1397

### CODE REVIEW

**Comment by schuetzm** [comment]
Can someone again explain what the actual problem is? I understand why you real default constructors are not possible, but why shouldn't the behaviour for constructors with default arguments be like this:

```D
struct S {
    int x;
    this(int x = 1) { this.x = x; }
}

S s1;
assert(s1.x == 0);
S s2 = S();
assert(s2.x == 1);
```

**Comment by yebblies** [comment]
Because default arguments do not have to be compile-time constants, so they can't be silently copied into the init value, and if you don't have them in the init value then it is a 'real' default ctor.

In some cases it could work, but I don't know what it gains us...

**Comment by schuetzm** [comment]
That's not what I mean. What you say applies to all constructors, those with and without default arguments. What is the problem with calling the constructor when the paren syntax is used, and using the .init value otherwise? Is there some non-obvious ambiguity?

File: src/func.d [hunk 1]
```diff
@@ -4437,20 +4437,50 @@ public:
...
         {
+            size_t dim = Parameter.dim(tf.parameters);
             StructDeclaration sd = ad.isStructDeclaration();
             if (sd)
             {
-                if (fbody || !(storage_class & STCdisable))
+                bool allDefaultedParams = (dim == 0) && tf.varargs == 1;  // c-style: foo(...)
+                bool isDefCtor = dim == 0;
+
+                for (size_t i = 0; i < dim; i++)
```
Comment (andralex) [inline src/func.d]: Since this is new code, it may as well be D-ified: `foreach (i; 0 .. dim)`

File: src/func.d [hunk 2]
```diff
@@ -4437,20 +4437,50 @@ public:
...
+
+                    // note: D-style variadic arguments allowed as a special-case (tf.varargs == 2)
+                    if (i + 1 == dim && arg.defaultArg)
+                    {
+                        isDefCtor = true;
+                        allDefaultedParams = true;
+                    }
+                    else
+                    {
+                        break;
```
Comment (andralex) [inline src/func.d]: Unnecessary control flow, should be:

```D
if (i + 1 != dim || !arg.defaultArg)
{
    break;
}
... rest of the code ...
```

**Comment by andralex** [comment]
lgtm with nits

**Comment by andralex** [comment]
Auto-merge toggled on

**Comment by yebblies** [comment]
> That's not what I mean. What you say applies to all constructors, those with and without default arguments. What is the problem with calling the constructor when the paren syntax is used, and using the .init value otherwise? Is there some non-obvious ambiguity?

Yeah, I don't know why we have `opCall` for that but can't use constructors.

**Comment by MartinNowak** [comment]
Thanks for your contribution but this is a breaking change w/o deprecation cycle, please turn this into a warning/deprecation/error first. Furthermore it should warn/error when declaring such a constructor, not when it gets used (same as for declaring a default constructor).
[Issue 15515 – default construction disabled for struct constructor with default arguments](https://issues.dlang.org/show_bug.cgi?id=15515)
If this doesn't get fixed soon, we'll have to revert it for 2.070.0.

**Comment by MartinNowak** [comment]
> Furthermore it should warn/error when declaring such a constructor, not when it gets used (same as for declaring a default constructor).

Or is this intended as a feature to support the common usage of ctors w/ default args for `auto s = S()`?

**Comment by AndrejMitrovic** [comment]
It is a warning? You need `-w` to trigger it.

**Comment by MartinNowak** [comment]
> It is a warning? You need -w to trigger it.

I see. The problem seems to be that the constructor gets `@disable`d which triggers an error on usage.
https://issues.dlang.org/show_bug.cgi?id=15500#c2
Maybe someone "tweaked" the code later on?

File: src/func.d (hunk 2 above)
Comment (MartinNowak) [inline src/func.d]: Why you need a loop at all if you're only looking at the last argument?

**Pull request merged**